core/src/main/java/org/apache/carbondata/core/carbon/datastore/AbstractBlockIndexStoreCache.java

```diff
@@ -0,0 +1,110 @@
+import org.apache.carbondata.core.carbon.datastore.block.TableBlockUniqueIdentifier;
+import org.apache.carbondata.core.carbon.metadata.blocklet.DataFileFooter;
+import org.apache.carbondata.core.util.CarbonUtil;
+import org.apache.carbondata.core.util.CarbonUtilException;
+
+/**
+ * This class validate and load the B-Tree in the executor lru cache
+ * @param <K> cache key
+ * @param <V> Block Meta data details
+ */
+public abstract class AbstractBlockIndexStoreCache<K, V>
+    implements Cache<TableBlockUniqueIdentifier, AbstractIndex> {
+  /**
+   * carbon store path
+   */
+  protected String carbonStorePath;
+  /**
+   * CarbonLRU cache
+   */
+  protected CarbonLRUCache lruCache;
+
+  /**
+   * table segment id vs blockInfo list
+   */
+  protected  Map<String, List<BlockInfo>> segmentIdToBlockListMap;
+
+
+  /**
+   * map of block info to lock object map, while loading the btree this will be filled
+   * and removed after loading the tree for that particular block info, this will be useful
+   * while loading the tree concurrently so only block level lock will be applied another
+   * block can be loaded concurrently
+   */
+  protected Map<BlockInfo, Object> blockInfoLock;
+
+  public AbstractBlockIndexStoreCache(String carbonStorePath, CarbonLRUCache lruCache) {
+    this.carbonStorePath = carbonStorePath;
+    this.lruCache = lruCache;
+    blockInfoLock = new ConcurrentHashMap<BlockInfo, Object>();
+    segmentIdToBlockListMap = new ConcurrentHashMap<>();
+  }
+
+  /**
+   * This method will get the value for the given key. If value does not exist
+   * for the given key, it will check and load the value.
+   *
+   * @param tableBlock
+   * @param tableBlockUniqueIdentifier
+   * @param lruCacheKey
+   */
+  protected void checkAndLoadTableBlocks(AbstractIndex tableBlock,
+      TableBlockUniqueIdentifier tableBlockUniqueIdentifier, String lruCacheKey)
+      throws CarbonUtilException {
+    // calculate the required size is
+    TableBlockInfo blockInfo = tableBlockUniqueIdentifier.getTableBlockInfo();
+    long requiredMetaSize = CarbonUtil
+        .calculateMetaSize(blockInfo.getFilePath(), blockInfo.getBlockOffset(),
+            blockInfo.getBlockLength());
+    if (requiredMetaSize > 0) {
+      tableBlock.setMemorySize(requiredMetaSize);
+      tableBlock.incrementAccessCount();
+      boolean isTableBlockAddedToLruCache = lruCache.put(lruCacheKey, tableBlock, requiredMetaSize);
+      // if column is successfully added to lru cache then only load the
+      // table blocks data
+      if (isTableBlockAddedToLruCache) {
+        // load table blocks data
+        // getting the data file meta data of the block
+        DataFileFooter footer = CarbonUtil
+            .readMetadatFile(blockInfo);
+        footer.setBlockInfo(new BlockInfo(blockInfo));
+        // building the block
+        tableBlock.buildIndex(Arrays.asList(footer));
+      } else {
+        throw new CarbonUtilException(
```

Comment:
Its bound to the Cache interface methods
  V get(K key) throws CarbonUtilException;
  List<V> getAll(List<K> keys) throws CarbonUtilException;